### PR TITLE
Refactors Card Labels with Custom Icons

### PR DESCRIPTION
**Pull request by redreceipt**
## DESCRIPTION

Screen Shot 2020-04-02 at 2 26 51 PM

### What does this PR do, or why is it needed?

Should have done this when I refactored `Chip`, this switches `CardLabels` to use the new API and allows custom icons.

### What design trade-offs/decisions were made?

None.

### How do I test this PR?

Pull up the stories.

### What automated tests did you write?

None.

## TODO:

- [x] PR has a relevant title that will be understandable in a public changelog (ie...non developers)
- [ ] Closes [tag-issues-here]
- [x] No new warnings in tests, in storybook, and in-app
- [ ] Upload GIF(s) of iOS and Android
- [x] Set a relevant reviewer

## REVIEW:

**Manual QA**

- [ ] Manual QA on iOS and ensure it looks/behaves as expected
- [ ] Manual QA on Android and ensure it looks/behaves as expected

**Code Review: Questions to consider**

- [ ] Read through the "Files changed" tab _very carefully_
- [ ] Edge cases: what assumptions are made about input?
- [ ] What kind of tests could be written?
- [ ] How might we make this easier for someone else to understand?
- [ ] Could the code be simpler?
- [ ] Will the code be easy to modify in the future?
- [ ] What's one part of these changes that makes you excited to merge it?

> The purpose of PR Review is to _improve the quality of the software._

### CODE REVIEW

File: packages/apollos-ui-kit/src/Card/Label.js [hunk 1]
```diff
@@ -16,14 +16,16 @@ const StyledChip = styled(
...
   <ThemeMixin
     mixin={{
       type: get(theme, 'type', 'light').toLowerCase(), // not sure why we need toLowerCase
       colors: get(theme, 'colors', {}),
     }}
   >
-    <StyledChip title={title} icon={icon} type={type} {...props} />
+    <StyledChip title={title} type={type} {...props}>
+      {icon ? <IconComponent name={icon} /> : null}
+    </StyledChip>
```
[the remark this answers is not in the document]
Comment (conrad-vanl) [reply packages/apollos-ui-kit/src/Card/Label.js:28]: > Can the question of a valid element not be handled solely by prop types? I thought that was its job.

The problem is that there's been two conventions used across the app, and we should standardize them. One convention is expecting a valid `Component Type` with props like this:

`<UIComponent MyCustomComponent={SomeComponentType} />`

The other convention is expecting a rendered React tree, which looks like this:

`<UIComponent MyCustomComponent={<SomeComponentType />} />`

For reference, any time `react-native` has `Component`-esque props, they support both use-cases. You can see the [exact same code I suggested above in their `ListView` component](https://github.com/facebook/react-native/blob/master/Libraries/Lists/VirtualizedList.js#L970-L975).

---

> `name={IconComponent ? null : icon}`

I'm not sure this is any better. `null` is still a value for `name` (`name={IconComponent ? undefined : icon}` would be better). But this would have the unintended consequence of overriding any `name` prop that was already passed to the underlying component. Say you did something like this:

```
MyCustomIconComponent = withTheme(({ theme }) => ({ name: theme.someThemeValue }))(Icon);

<CardLabel IconComponent={MyCustomIconComponent} />
```

The `MyCustomIconComponent` would never actually get `theme.someThemeValue` for the `name` prop, since you are passing in `name={null}` (or even `name={undefined}` would clear the prop out).

---

So I think the solution here:

1. Shouldn't make assumptions of how `IconComponent` is used.
2. Passing in `name={icon}` isn't a problem as long as it's consistent and documented. Selectively `null`-ing it or making it `undefined` introduces debt
3. We should follow conventions established by `react-native` for props and what they do based on their name. If `react-native` has a convention of supporting both `PropTypes.elementType` and `PropTypes.element` for props that have `Component` in their name, we should do the same here.

[the remark this answers is not in the document]
Comment (redreceipt) [reply packages/apollos-ui-kit/src/Card/Label.js:28]: Sounds good. Just so I'm clear, what do you think should happen with the four possible combinations of `icon` and `IconComponent`?

[the remark this answers is not in the document]
Comment (redreceipt) [reply packages/apollos-ui-kit/src/Card/Label.js:28]: Actually, to be more specific, what do you think should happen when both are passed? The other cases are straightforward.

[the remark this answers is not in the document]
Comment (redreceipt) [reply packages/apollos-ui-kit/src/Card/Label.js:28]: Oh, I see your suggestion, so something like this?

```
    <StyledChip title={title} type={type} {...props}>
      {icon || IconComponent ?
         <IconComponent {...(IconComponent ? {} : { name: icon } )} />
      ) : undefined}
    </StyledChip>
```

[the remark this answers is not in the document]
Comment (conrad-vanl) [reply packages/apollos-ui-kit/src/Card/Label.js:28]: @redreceipt I'm fine with always passing the `name` prop into `IconComponent` if it's given to us. That way there's fewer 'special cases' going on.

So - 

- `IconComponent` == `<IconComponent name={icon} />`
- `<IconComponent />` == `<IconComponent />` (if an element is passed, just render it)
- `icon` and `!IconComponent` === `<defaultProps.IconComponent name={icon} />`
- `!icon` and `!IconComponent` === nothing

[the remark this answers is not in the document]
Comment (conrad-vanl) [reply packages/apollos-ui-kit/src/Card/Label.js:28]: @redreceipt actually I think just this is fine - works for all cases

```suggestion
    <StyledChip title={title} type={type} {...props}>
      {IconComponent ? (React.isValidElement(IconComponent) ? (
        IconComponent
      ) : (
        <IconComponent name={icon} />
      )) : undefined}
    </StyledChip>
```

[the remark this answers is not in the document]
Comment (redreceipt) [reply packages/apollos-ui-kit/src/Card/Label.js:28]: that doesn't allow for only passing `icon`

[the remark this answers is not in the document]
Comment (redreceipt) [reply packages/apollos-ui-kit/src/Card/Label.js:28]: what about this

Screen Shot 2020-04-06 at 2 34 39 PM

[the remark this answers is not in the document]
Comment (redreceipt) [reply packages/apollos-ui-kit/src/Card/Label.js:28]: actually, couldn't we just handle a nameless icon in the `Icon` component itself (_might already_)? So this could just be:

Screen Shot 2020-04-06 at 2 42 51 PM

**Comment by redreceipt** [comment]
@conrad-vanl ok so this latest solution just lets `Chip` handle the deprecated `icon` prop. I think it's pretty clean.
